<launch>

    <!--arg name="tf_world"  default="0.0 0.0 0.0 0.0 0.0 0.0"/-->
    <arg name="tf_world"  default="40.0 -106.5 -0.32 0.0 0.0 110.5"/>

	
    <rosparam file = "$(find espeleo_locomotion)/config/locomotion_parameters.yaml"
              command = "load"
              ns = "">
    </rosparam>


    <node name = "espeleo_locomotion_wheels"
          pkg = "espeleo_locomotion"
          type = "espeleo_locomotion_differential.py"
          respawn = "true"
          output="screen">
    </node>


    <node name = "espeleo"
          pkg = "espeleo_bringup"
          type = "dyn_reconfigure_server.py"
          respawn = "true"
          output = "screen">
    </node>


    <node name = "rgb_flip"
          pkg = "espeleo_mythings"
          type = "flip_image.py"
          output = "screen">
          <remap from = "/image_raw_flipped"   to = "/d435/rgb/image_raw_flipped"/>
          <remap from = "/camera_info_flipped" to = "/d435/rgb/camera_info"/>
          <remap from = "/image_raw"           to = "/rgb/image_rect"/>
          <remap from = "/camera_info"         to = "/rgb/camera_info"/>
    </node>


    <node name = "depth_flip"
          pkg = "espeleo_mythings"
          type = "flip_image.py"
          output = "screen">
          <remap from = "/image_raw_flipped"   to = "/d435/depth/image_rect_flipped"/>
          <remap from = "/camera_info_flipped" to = "/d435/depth/camera_info"/>
          <remap from = "/image_raw"           to = "/depth/image_rect"/>
          <remap from = "/camera_info"         to = "/depth/camera_info"/>
    </node>


    <node name = "tfs_publisher"
          pkg  = "espeleo_mythings"
          type = "tower_tfs.py"
          args = "_tf_world:='$(arg tf_world)'"
          output = "screen">
    </node>


    <node name   = "odom_to_init_node"
          pkg    = "espeleo_mythings"
          type   = "odom_to_init.py"
          args   = "_tf_world:='$(arg tf_world)'"
          output = "screen">
    </node>


    <node name   = "rtab_imu_node"
          pkg    = "espeleo_mythings"
          type   = "imu_repub.py"
          args   = "_frame_id:='rtab_imu'"
          output = "screen">
    </node>


    <include file = "$(find rtabmap_ros)/launch/rtabmap.launch">

        <arg name = "rgb_topic"               value = "/rgb/image_rect"/>
        <arg name = "depth_topic"             value = "/depth/image_rect"/>
        <arg name = "camera_info_topic"       value = "/rgb/camera_info"/>
        <arg name = "depth_camera_info_topic" value = "/depth/camera_info"/>

        <arg name = "args"                    value = "--delete_db_on_start
                                                       --Optimizer/GravitySigma 0.3
                                                       --Odom/Holonomic false
                                                       --Odom/ResetCountdown 1
                                                       --Rtabmap/StartNewMapOnLoopClosure true
                                                       --Odom/Strategy 1
                                                       --Reg/Force3DoF false"/>

        <arg name = "queue_size"              value = "10"/>
        <arg name = "approx_sync"             value = "true"/>
        <arg name = "wait_imu_to_init"        value = "false"/>
        <arg name = "visual_odometry"         value = "true"/>

        <arg name = "odom_topic"              value = "/rtabmap/odom"/>
        <arg name = "imu_topic"               value = "/imu/data2x"/>
        <arg name = "map_frame_id"            value = "rtab_init"/>
        <arg name = "frame_id"                value = "rtab_pose"/>
        <arg name = "use_sim_time"            value = "false"/>
        <arg name = "rtabmapviz"              value = "true"/>
        <arg name = "rviz"                    value = "false"/>

    </include>


    <node name   = "odom_node"
          pkg    = "espeleo_mythings"
          type   = "odom.py"
          args   = ""
          output = "screen">
    </node>


    <node name   = "ekf_imu_node"
          pkg    = "espeleo_mythings"
          type   = "imu_repub.py"
          args   = "_frame_id:='ekf_imu'"
          output = "screen">
          <remap from = "/imu/data2"
                 to = "/imu/data3"/>
    </node>


    <node name   = "espeleo_pose_ekf"
          pkg    = "espeleo_pose_ekf"
          type   = "espeleo_pose_ekf"
          args   = ""
          output = "screen">

        <remap from = "/odom"                      to = "/robot_odom"/>
        <remap from = "/imu_data"                  to = "/imu/data3"/>
        <remap from = "/espeleo_pose_ekf/ekf_odom" to = "/ekf/odom"/>

    </node>


</launch>
